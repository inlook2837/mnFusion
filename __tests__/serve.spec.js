jest.setTimeout(100000)

const create = require('@vue/cli-test-utils/createTestProject')
const path = require('path')
const Application = require('spectron').Application
const electronPath = require('electron')
const { defaultPreset } = require('@vue/cli/lib/options')

const serve = (project, notifyUpdate) =>
  new Promise((resolve, reject) => {
    const child = project.run('vue-cli-service serve:electron')
    let isFirstMatch = true
    let log = ''
    child.stdout.on('data', async data => {
      data = data.toString()
      log += data
      try {
        const urlMatch = data.match(/http:\/\/[^/]+\//)
        if (urlMatch && isFirstMatch) {
          isFirstMatch = false
          let url = urlMatch[0]

          resolve({
            stdout: log,
            url,
            stopServe: () => {
              child.stdin.write('close')
            }
          })
        } else if (data.match(/App updated/)) {
          if (notifyUpdate) {
            notifyUpdate(data)
          }
        } else if (data.match(/Failed to compile/)) {
          reject(data)
        }
      } catch (err) {
        reject(err)
      }
    })
  })
const runTests = useTS =>
  new Promise(async resolve => {
    let preset = { ...defaultPreset }
    let projectName = 'serve'
    const projectPath = p =>
      path.join(
        process.cwd(),
        '__tests__/projects/serve' + (useTS ? '-ts' : ''),
        p
      )
    if (useTS) {
      defaultPreset.plugins['@vue/cli-plugin-typescript'] = {}
      projectName += '-ts'
    }
    // Install vcp-electron-builder
    defaultPreset.plugins['vue-cli-plugin-electron-builder'] = {}
    const project = await create(
      projectName,
      preset,
      path.join(process.cwd(), '/__tests__/projects')
    )
    //   Prevent electron from being launched
    jest.mock('execa')
    //   Wait for dev server to start
    const { stopServe, url } = await serve(project)
    expect(project.has('dist_electron/background.js')).toBe(true)
    const app = new Application({
      path: electronPath,
      args: [projectPath('dist_electron/background.js')],
      env: {
        WEBPACK_DEV_SERVER_URL: url,
        IS_TEST: true
      },
      cwd: projectPath('')
    })

    await app.start()
    const win = app.browserWindow
    const client = app.client
    await client.waitUntilWindowLoaded()

    await client.getRenderProcessLogs().then(logs => {
      logs.forEach(log => {
        //   Make sure there are no fatal errors
        expect(log.level).not.toBe('SEVERE')
      })
    })
    await client.getMainProcessLogs().then(logs => {
      logs.forEach(log => {
        //   Make sure there are no fatal errors
        expect(log.level).not.toBe('SEVERE')
      })
    })
    //   Window was created
    expect(await client.getWindowCount()).toBe(1)
    //   It is not minimized
    expect(await win.isMinimized()).toBe(false)
    //   Window is visible
    expect(await win.isVisible()).toBe(true)
    //   Size is correct
    const { width, height } = await win.getBounds()
    expect(width).toBeGreaterThan(0)
    expect(height).toBeGreaterThan(0)
    //   App is loaded properly
    expect(await client.getHTML('#app')).toMatchSnapshot()

    stopServe()
    await app.stop()
    resolve()
  })
test('serve:electron', async () => {
  await runTests()
})
test('serve:electron-ts', async () => {
  await runTests(true)
})
