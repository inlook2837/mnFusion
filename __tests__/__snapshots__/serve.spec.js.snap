// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`serve:electron 1`] = `
"<div id=\\"app\\"><img src=\\"/img/logo.82b9c7a5.png\\"><div data-v-656039f0=\\"\\" class=\\"hello\\"><h1 data-v-656039f0=\\"\\">Welcome to Your Vue.js App</h1><p data-v-656039f0=\\"\\">
    For guide and recipes on how to configure / customize this project,<br data-v-656039f0=\\"\\">
    check out the
    <a data-v-656039f0=\\"\\" href=\\"https://cli.vuejs.org\\" target=\\"_blank\\">vue-cli documentation</a>.
  </p><h3 data-v-656039f0=\\"\\">Installed CLI Plugins</h3><ul data-v-656039f0=\\"\\"><li data-v-656039f0=\\"\\"><a data-v-656039f0=\\"\\" href=\\"https://github.com/vuejs/vue-cli/tree/dev/packages/%40vue/cli-plugin-babel\\" target=\\"_blank\\">babel</a></li><li data-v-656039f0=\\"\\"><a data-v-656039f0=\\"\\" href=\\"https://github.com/vuejs/vue-cli/tree/dev/packages/%40vue/cli-plugin-eslint\\" target=\\"_blank\\">eslint</a></li><li data-v-656039f0=\\"\\"><a data-v-656039f0=\\"\\" href=\\"https://www.npmjs.com/package/vue-cli-plugin-electron-builder\\" target=\\"_blank\\">electron-builder</a></li></ul><h3 data-v-656039f0=\\"\\">Essential Links</h3><ul data-v-656039f0=\\"\\"><li data-v-656039f0=\\"\\"><a data-v-656039f0=\\"\\" href=\\"https://vuejs.org\\" target=\\"_blank\\">Core Docs</a></li><li data-v-656039f0=\\"\\"><a data-v-656039f0=\\"\\" href=\\"https://forum.vuejs.org\\" target=\\"_blank\\">Forum</a></li><li data-v-656039f0=\\"\\"><a data-v-656039f0=\\"\\" href=\\"https://chat.vuejs.org\\" target=\\"_blank\\">Community Chat</a></li><li data-v-656039f0=\\"\\"><a data-v-656039f0=\\"\\" href=\\"https://twitter.com/vuejs\\" target=\\"_blank\\">Twitter</a></li></ul><h3 data-v-656039f0=\\"\\">Ecosystem</h3><ul data-v-656039f0=\\"\\"><li data-v-656039f0=\\"\\"><a data-v-656039f0=\\"\\" href=\\"https://router.vuejs.org\\" target=\\"_blank\\">vue-router</a></li><li data-v-656039f0=\\"\\"><a data-v-656039f0=\\"\\" href=\\"https://vuex.vuejs.org\\" target=\\"_blank\\">vuex</a></li><li data-v-656039f0=\\"\\"><a data-v-656039f0=\\"\\" href=\\"https://github.com/vuejs/vue-devtools#vue-devtools\\" target=\\"_blank\\">vue-devtools</a></li><li data-v-656039f0=\\"\\"><a data-v-656039f0=\\"\\" href=\\"https://vue-loader.vuejs.org\\" target=\\"_blank\\">vue-loader</a></li><li data-v-656039f0=\\"\\"><a data-v-656039f0=\\"\\" href=\\"https://github.com/vuejs/awesome-vue\\" target=\\"_blank\\">awesome-vue</a></li></ul></div></div>"
`;
